### ISSUE TITLE
Filters are not cached when both cache_index_and_filter_blocks and allow_mmap_reads are set

### ISSUE THREAD

**Issue by kobykahane** [facebook/rocksdb#9367]
After attempting to upgrade from an older version of RocksDB to a newer one that includes  #5298 and #5504, I started seeing a very significant slowdown in a point-lookup heavy workload. Profiling indicated that filter blocks are repeatedly being read and checksummed (`rocksdb::FilterBlockReaderCommon<rocksdb::ParsedFullFilterBlock>::GetOrReadFilterBlock` invoking `rocksdb::BlockFetcher::ReadBlockContents` and `rocksdb::VerifyBlockChecksum` constantly, when they were not being invoked before).

The following test program demonstrates the issue. It performs a 100 Puts and then performs 100 Gets with a perf context. Note `cache_index_and_filter_blocks` and `allow_mmap_reads` are set. When executed with RocksDB 6.4 and later, 100 filter block reads occur:

```c++
#include <cstdlib>
#include <cassert>
#include <iostream>
#include <memory>
#include <string>
#include <iostream>

#include <rocksdb/db.h>
#include <rocksdb/table.h>
#include <rocksdb/options.h>
#include <rocksdb/filter_policy.h>
#include <rocksdb/cache.h>
#include <rocksdb/perf_context.h>
#include <rocksdb/perf_level.h>

static rocksdb::Status openDB(rocksdb::Options const& opts, std::string const& name, std::unique_ptr<rocksdb::DB>& db)
{
    rocksdb::DB* pdb{};
    rocksdb::Status s = rocksdb::DB::Open(opts, name, &pdb);
    if (s.ok()) {
        db.reset(pdb);
    }

    return s;
}

static void check_status(rocksdb::Status s)
{
    if (!s.ok()) {
        std::cerr << s.ToString() << '\n';
        abort();
    }
}

int main()
{
    rocksdb::Options opts;
    opts.create_if_missing = true;
    opts.statistics = rocksdb::CreateDBStatistics();
    opts.compression = rocksdb::kLZ4Compression;
    opts.allow_mmap_reads = true;

    rocksdb::BlockBasedTableOptions bbto;
    bbto.cache_index_and_filter_blocks = true;
    bbto.block_cache = rocksdb::NewLRUCache(32 * 1024 * 1024);
    bbto.filter_policy.reset(rocksdb::NewBloomFilterPolicy(10, false));

    opts.table_factory.reset(rocksdb::NewBlockBasedTableFactory(bbto));

    rocksdb::DestroyDB("rdb_test", opts);
    std::unique_ptr<rocksdb::DB> db;
    rocksdb::Status s = openDB(opts, "rdb_test", db);
    check_status(s);

    constexpr int keyCount = 100;

    rocksdb::WriteBatch wb;
    for (int i = 0; i < keyCount; i++) {
        auto kv = std::to_string(i);
        s = wb.Put(kv, kv);
        check_status(s);
    }

    rocksdb::WriteOptions wo;
    s = db->Write(wo, &wb);
    check_status(s);

    rocksdb::FlushOptions fo;
    s = db->Flush(fo);
    check_status(s);

    rocksdb::ReadOptions ro;

    rocksdb::SetPerfLevel(rocksdb::PerfLevel::kEnableCount);
    auto pc = rocksdb::get_perf_context();
    pc->EnablePerLevelPerfContext();
    pc->Reset();

    std::string vd;
    for (int i = 0; i < keyCount; i++) {
        auto kv = std::to_string(i);
        s = db->Get(ro, kv, &vd);
        if (!s.ok() && !s.IsNotFound()) {
            abort();
        }
    }

    std::cout << "Filter block read count: " << pc->filter_block_read_count << '\n';
    pc->Reset();

    s = db->Close();
    check_status(s);

    return 0;
}
```

With RocksDB 6.4 and higher the output is:
```
Filter block read count: 100
```
With RocksDB 6.3.6 and earlier, or with RocksDB 6.4 and higher with either `allow_mmap_reads = false` or `cache_index_and_filter_blocks = false`:
```
Filter block read count: 0
```

### Expected behavior
Filter blocks should be cached and not read and checksummed repeatedly when `cache_index_and_filter_blocks` and `allow_mmap_reads` are both set.
### Actual behavior
If both `cache_index_and_filter_blocks` and `allow_mmap_reads` are set, filter blocks are read repeatedly.
### Steps to reproduce the behavior
See above program.

@ltamasi - FYI.

**Comment by ltamasi** [comment]
@kobykahane In general, blocks read from mmapped files are not safe to put in the block cache because the block object doesn't own the byte range in this case (see the relevant check [here](https://github.com/facebook/rocksdb/blob/9a116ab4b4af980132c2deb4d6a68a9687030b14/table/block_based/block_based_table_reader.cc#L1329)). In particular, inserting an mmapped block into the cache can leave a dangling pointer to the mmap area if the underlying file is closed. This is something that had been broken previously. I would suggest setting `cache_index_and_filter_blocks` to `false` when using mmapped I/O.

**Comment by kobykahane** [comment]
> @kobykahane In general, blocks read from mmapped files are not safe to put in the block cache because the block object doesn't own the byte range in this case (see the relevant check [here](https://github.com/facebook/rocksdb/blob/9a116ab4b4af980132c2deb4d6a68a9687030b14/table/block_based/block_based_table_reader.cc#L1329)). In particular, inserting an mmapped block into the cache can leave a dangling pointer to the mmap area if the underlying file is closed. This is something that had been broken previously. I would suggest setting `cache_index_and_filter_blocks` to `false` when using mmapped I/O.

Thanks. We use `cache_index_and_filter_blocks` to have bounded filter/index block memory usage. I suppose the implication is that we should not use mmapped I/O. I think the comments for `allow_mmap_reads` should be updated to warn that it essentially disables caching for index and filter blocks when they are not eagerly loaded by table readers.

Additionally, do I understand correctly that if `partition_filters` is enabled, even with `cache_index_and_filter_blocks = false`, if `allow_mmap_reads` is set, partitioned filter blocks would never be cached? So mmapped I/O shouldn't be used with partitioned filters either?

**Comment by ltamasi** [comment]
Right, the above point about the blocks of mmapped files being cache-unsafe holds for filter partitions as well. On the other hand, reading from a paged-in region of a memory-mapped file should be pretty fast.

**Comment by kobykahane** [comment]
> Right, the above point about the blocks of mmapped files being cache-unsafe holds for filter partitions as well. On the other hand, reading from a paged-in region of a memory-mapped file should be pretty fast.

The read itself appears to be fast, but the repeated calls to VerifyBlockChecksum after every such read seem to dominate CPU time during the repeated point lookup scenario.

**Comment by ajkr** [comment]
Should we copy the data into block cache when mmap reads and block cache are both enabled? I imagine not doing so can also cause uneven query perf when a subset of data blocks are stored uncompressed due to poor compression ratio.

I think #5504 surfaced this by fixing an issue (previously with mmap read, querying a filter block could transparently read from storage without verifying checksum).

**Comment by ltamasi** [comment]
> Should we copy the data into block cache when mmap reads and block cache are both enabled?

Yeah, that's what I was thinking too. 

> I imagine not doing so can also cause uneven query perf when a subset of data blocks are stored uncompressed due to poor compression ratio.

Right. In addition, index blocks might be uncompressed as well.
